source/loader/layers/validation/ur_leak_check.hpp

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: MIT
+#ifndef UR_LEAK_CHECK_H
+#define UR_LEAK_CHECK_H 1
+
+#include "backtrace.hpp"
+#include "ur_validation_layer.hpp"
+
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+
+#define MAX_BACKTRACE_FRAMES 64
+
+namespace validation_layer {
+struct RefCountContext {
+  private:
+    struct RefRuntimeInfo {
+        int64_t refCount;
+        std::vector<BacktraceLine> backtrace;
+    };
+
+    std::mutex mutex;
+    std::unordered_map<void *, struct RefRuntimeInfo> counts;
+
+    inline struct RefRuntimeInfo createRefInfo(int refCount) {
+        std::vector<BacktraceLine> backtrace = getCurrentBacktrace();
+
+        return RefRuntimeInfo{refCount, backtrace};
+    }
+
+    std::pair<void *, int64_t> incrementRefCount(void *ptr) {
+        auto it = counts.find(ptr);
+        if (it != counts.end()) {
+            it->second.refCount++;
+            return std::make_pair(it->first, it->second.refCount);
+        } else {
+            counts.emplace(ptr, createRefInfo(1));
+            return std::make_pair(ptr, 1);
+        }
+    }
+
+    std::pair<void *, int64_t> decrementRefCount(void *ptr) {
+        auto it = counts.find(ptr);
+        if (it != counts.end()) {
+            it->second.refCount--;
+            return std::make_pair(it->first, it->second.refCount);
+        } else {
+            counts.emplace(ptr, createRefInfo(-1));
```

Comment:
isn't this an error anyway? you can't decrement something that doesn't exist. We should log this as an error.

Reply:
There were suggestions that tracking *Release calls that have gone into the negative could be useful when the UR API wouldn't throw an error for some reason.

Reply:
My suggestion was to differentiate between the scenario where retain is called on a completely invalid object (e.g., `0xCAFE`) and where retain is called on a previously valid one (going into negative).

Reply:
As we discussed, to differentiate between invalid references and valid references that have gone into negative we need to keep a record of all involved references. The memory cost is probably not worth the effort.